llvm/include/llvm/Object/OffloadBinary.h

```diff
@@ -183,11 +210,166 @@ class OffloadFile : public OwningBinary<OffloadBinary> {
   }
 };
 
+/// Bundle entry in binary clang-offload-bundler format.
+struct OffloadBundleEntry {
+  uint64_t Offset = 0u;
+  uint64_t Size = 0u;
+  uint64_t IDLength = 0u;
+  StringRef ID;
+  OffloadBundleEntry(uint64_t O, uint64_t S, uint64_t I, StringRef T)
+      : Offset(O), Size(S), IDLength(I), ID(T) {}
+  void dumpInfo(raw_ostream &OS) {
+    OS << "Offset = " << Offset << ", Size = " << Size
+       << ", ID Length = " << IDLength << ", ID = " << ID;
+  }
+  void dumpURI(raw_ostream &OS, StringRef FilePath) {
+    OS << ID.data() << "\tfile://" << FilePath << "#offset=" << Offset
+       << "&size=" << Size << "\n";
+  }
+};
+
+/// Fat binary embedded in object files in clang-offload-bundler format
+class OffloadBundleFatBin {
+
+  uint64_t Size = 0u;
+  StringRef FileName;
+  uint64_t NumberOfEntries;
+  SmallVector<OffloadBundleEntry> Entries;
+
+public:
+  SmallVector<OffloadBundleEntry> getEntries() { return Entries; }
+  uint64_t getSize() const { return Size; }
+  StringRef getFileName() const { return FileName; }
+  uint64_t getNumEntries() const { return NumberOfEntries; }
+
+  static Expected<std::unique_ptr<OffloadBundleFatBin>>
+  create(MemoryBufferRef, uint64_t SectionOffset, StringRef FileName);
+  Error extractBundle(const ObjectFile &Source);
+
+  Error dumpEntryToCodeObject();
+
+  Error readEntries(StringRef Section, uint64_t SectionOffset);
+  void dumpEntries() {
+    SmallVectorImpl<OffloadBundleEntry>::iterator It = Entries.begin();
+    for (uint64_t I = 0; I < Entries.size(); I++) {
+      It->dumpInfo(outs());
+      ++It;
+    }
+  }
+
+  void printEntriesAsURI() {
+    SmallVectorImpl<OffloadBundleEntry>::iterator It = Entries.begin();
+    for (uint64_t I = 0; I < Entries.size(); I++) {
+      It->dumpURI(outs(), FileName);
+      ++It;
+    }
```

Comment:
```suggestion
  for (OffloadBundleEntry &Entry : Entries)
    Entry.dumpURI(outs(), Filename);
```
And everywhere else you could trivially replace the for-loop.